Initial Commit of dagster-cloud.yaml doc (#13533)
dagster-io/dagster: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx

---
title: Confiuguration options for dagster_cloud.yaml | Dagster Docs
description: How to use dagster_cloud.yaml to configure Dagster Cloud
---

# Dagster_cloud.yaml Guide

<Note>This guide is applicable to Dagster Cloud.</Note>

This dagster_cloud.yaml file is used to configure a Dagster Cloud instance. The file is required to tell Dagster where the code sources are to load into Dagster Cloud but has other options as well. For Hybrid Deployments, secrets are configured in this file as well.

## Where to put the dagster_cloud.yaml file

The dagster_cloud.yaml file should be placed in your dagster project's repository root. Below is an example of a file structure from the [Dagster cloud ETL quickstart](https://github.com/dagster-io/dagster/tree/master/examples/quickstart_etl).

```shell
quickstart_etl
├── README.md
├── quickstart_etl
│   ├── __init__.py
│   ├── assets
├── quickstart_etl_tests
├── dagster_cloud.yaml
├── pyproject.toml
├── setup.cfg
└── setup.py
```

## Code Source

> The only **mandatory** section in the dagster_cloud.yaml file is setting up your code locations. If you have multiple code locations, you will have multiple code locations within a single dagster_cloud.yaml file. Dagster code can be sourced from a package, a python file, or a module.

Reviewer: ```suggestion
The only **mandatory** section in the dagster_cloud.yaml file is setting up your code locations. If you have multiple code locations, you will have multiple code locations within a single dagster_cloud.yaml file. Dagster code can be sourced from a package, a python file, or a module, and one of these parameters should be specified for each code location.
```

Parameter reference table:

| Parameter                            | Description                                                                           | Data type            |
| ------------------------------------ | ------------------------------------------------------------------------------------- | -------------------- |
| `locations.location_name`            | The name of your code location that will appear in the Dagster UI code locations page | string               |

> | `locations.code_source.package_name` | The folder that contains your dagster code                                            | string (folder name) |

Reviewer: ```suggestion
| `locations.code_source.package_name` |  Use if your dagster code is structured as a package | string (folder name) |
```

> | `locations.code_source.python_file`  | The python file that contains your dagster code                                       | string (.py file)    |

Reviewer: ```suggestion
| `locations.code_source.python_file`  | Use if your dagster code is in a single file | string (.py file)    |
```

| `locations.code_source.module_name`  | The module that contains your dagster code                                            | string (module name) |

> Single code location:

Reviewer: ```suggestion
Example for a single code location:
```

I'd add the example pre-amble throughout

```yaml
#dagster_cloud.yaml

locations:
  - location_name: example_location
    code_source:
      package_name: my_dagster_project

>     build:

Reviewer: We should remove the build section here because it is not explained until later in the doc.

      directory: ./
      registry: 764506304434.dkr.ecr.us-west-2.amazonaws.com/branch-deployments-gh-action-test
```

Multiple code locations:

```yaml
#dagster_cloud.yaml

locations:
  - location_name: data-eng-pipeline
    code_source:
      package_name: hooli_data_eng
  - location_name: basics
    code_source:
      python_file: definitions.py
```

## Configuring your working directory

There might be some cases where you want to run your Dagster code from a different directory than the root of your code repository. You can specify the directory you want to run your Dagster code from in the dagster_cloud.yaml file using the `working_directory` option.

Parameter reference table:

| Parameter           | Description                                            | Data type     |
| ------------------- | ------------------------------------------------------ | ------------- |
| `working_directory` | The path of the directory that Dagster should run from | string (path) |

```yaml
#dagster_cloud.yaml

locations:
  - location_name: data-eng-pipeline
    code_source:
      package_name: hooli_data_eng

>     working_directory: ./hooli_data_eng

Reviewer 2: I suggest a different directory name because this field is unrelated to `package_name`.

```

## Build: Docker Registry

> 

Reviewer: We should add a pre-amble here.

If you are using Dagster Cloud Hybrid with a Docker, Kubernetes, or ECR agent you will need to tell Dagster Cloud where the Docker image containing your code is stored. This is done through the`build` section of dagster_cloud.yaml.

Reviewer: @gibsondan is this all true? I'm unclear as to what situations require `build` vs just including the `image` value. I think it has to do with whether you use our build_and_deploy action or just deploy... Any ideas on what language could make that distinction clear here?

Reviewer 2: This section should just be called the "Build directory". You can use a `build: directory:` in serverless as well. It works for both docker and pex. Basically it specifies which directory contains your setup.py or requirements.txt file for this code location.

Reviewer 2: We should say this is useful if your setup.py or requirements.txt is not at your project root but in a subdirectory. "This is common if you have multiple python packages within a single dagster project"

Parameter reference table:

| Parameter                   | Description                                                                                                                                                       | Data type                |
| --------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------ |

> | `locations.build.directory` | The directory in your project code that you want to deploy, the default is `.`. If you have sub-folders you can specific the path to only deploy that sub-project | string (path)            |

Reviewer: ```suggestion
| `locations.build.directory` | Use if you are using the hybrid cloud deploy GitHub action that builds a Docker image for you. Supply the directory in your project code that you want to include in the image, the default is `.`. If you have sub-folders you can specify the path to only include that sub-directory | string (relative path)            |
```

Reviewer 2: This doesn't apply only to hybrid, see my comment above.

> | `locations.build.registry`  | The Docker registry to push your code location to in the registry                                                                                                 | string (docker registry) |

Reviewer 2: Registry is hybrid only.

Single code location:

```yaml
#dagster_cloud.yaml

locations:
  - location_name: example_location
    code_source:
      package_name: my_dagster_project
    build:
      directory: ./

>       registry: 764506304434.dkr.ecr.us-west-2.amazonaws.com/branch-deployments-gh-action-test

Reviewer: We should make this registry reference, and all subsequent, more clearly a thing the customer needs to provide to avoid them copying and pasting this value

Maybe `registry: your-docker-image-registry/image-name`

```

> Multiple code locations:

Reviewer: I think this is not needed since we covered how to do multiple code locations earlier, and this section just focuses on the new build section

```yaml
#dagster_cloud.yaml

locations:
  - location_name: data-eng-pipeline
    code_source:
      package_name: hooli_data_eng
    build:
      directory: ./
      registry: 764506304434.dkr.ecr.us-west-2.amazonaws.com/hooli-data-science-prod
  - location_name: basics
    code_source:
      python_file: definitions.py
    build:
      directory: ./hooli_basics
      registry: 764506304434.dkr.ecr.us-west-2.amazonaws.com/hooli-data-science-prod
```

## Python Executable

> By default, Dagster Cloud will use the python executable that is installed on the agent. If you want to use a different python executable, you can specify it in the dagster_cloud.yaml file. It might also be useful to have different python executables for different code locations.

Reviewer: ```suggestion
If you are using Dagster Cloud Serverless, the Python executable to use is specified [here](https://docs.dagster.io/dagster-cloud/deployment/serverless#using-a-different-python-version). For Dagster Cloud Hybrid, the python executable that is installed globally in the image, or the default Python executable on the local system if you use the local agent, will be used. If you want to use a different python executable, you can specify it in the dagster_cloud.yaml file. It might also be useful to have different python executables for different code locations.
```

Parameter reference table:

| Parameter                   | Description                            | Data type     |
| --------------------------- | -------------------------------------- | ------------- |
| `locations.executable_path` | The location of your python executable | string (path) |

```yaml
#dagster_cloud.yaml

locations:
  - location_name: data-eng-pipeline
    code_source:

>       package_name: hooli_data_eng

Reviewer: location name and package name could reflect the executable path, eg data-eng stuff vs ml stuff

    executable_path: venvs/path/to/dataengineering_spark_team/bin/python
  - location_name: basics
    code_source:
      python_file: definitions.py
    executable_path: venvs/path/to/ml_tensorflow/bin/python
```

## Other Configurations

> 

Reviewer: I would ditch the section on secrets here. It is true that you can set secrets in some Dagster Cloud Hybrid setups, but many customers will still use the Cloud environment variable workflow. Instead I would say this: 

In addition to these settings, different Dagster Cloud Hybrid agents support additional configuration options. To see details: 

- [ECS](https://docs.dagster.io/dagster-cloud/deployment/agents/amazon-ecs/configuration-reference#per-location-configuration)
- [Kubernetes](https://docs.dagster.io/dagster-cloud/deployment/agents/kubernetes/configuration-reference#per-location-configuration)
- [Docker](https://docs.dagster.io/dagster-cloud/deployment/agents/docker/configuration-reference#environment-variables-and-secrets)
